### DISCUSSION TITLE
Graceful Shutdown through extensions

### DISCUSSION THREAD

**Discussion by flevi29** [directus/directus#13731] [Archive]
I am running Directus as a `worker_thread` `Worker`. For some reason when I try to shut down Directus with `process.emit('SIGINT')` within the thread, it itself shuts down gracefully but crashes my main thread with an exit code of 1, without even triggering `process.on('exit', ...)`. Something weird that terminus does must be the culprit, I believe.

Another option that I found is for me to shut it down manually (not ideal), but even this has a major blockade, because I need an instance of the server object for whatever reason for this to work.

```ts
        // within defineHook callback somewhere
        async stop() {
          await database.destroy();
          logger.info('Database connections destroyed');
          emitter.emitAction(
            'server.stop',
            // how do I get the server instance
            { server: ??? },
            {
              database,
              schema: null,
              accountability: null,
            },
          );
        }
```

I thought maybe through `action('server.start', (server) => ...)` I could get the instance, but that's just metadata, or so it seems, either way it doesn't work, and I'm not even sure if this whole thing could work.

So I'm asking is there a way to make this work? Or, far better, is there a one function shut down? Is there any chance one will be implemented some day?

Right now my only two choices are to either gracefully shut down Directus or my main app.

**Comment by flevi29** [comment]
Actually with this I can gracefully destroy database connections, and then just terminate Directus from main thread. Does this leave any vulnerabilities/data corruption risks? If not then it's probably good enough.

**Comment by rijkvanzanten** [comment]
No clue! I've never tried starting/stopping Directus from a worker thread, so I can't comment on the usability / success of any of these methods. That being said, you could also consider importing the Express app instance from `directus/app` rather than starting the server through a process 🤔 That way it might be a tad easier to control when to start/stop the webserver and/or inject it under your own app assuming it's express middleware compatible

**Reply by flevi29** [reply]
I'm not sure if integrating Directus app instance with my main app would work or if it even would be a good idea. `directus/server` also exports `createServer` that does a lot of seemingly important things and returns the server as well, so that might be better to fiddle with. `startServer` seems to do only basic starting/logging/error handling and a seemingly important `emitAction('server.start', ...)`, so overriding that function with my own slightly different one doesn't seem like it'll ever break in case of updates.

Either way for now I believe having the ability to destroy database connections is good enough.
